flink-connectors/flink-connector-files/src/main/java/org/apache/flink/connector/file/src/impl/DynamicFileSplitEnumerator.java

```diff
@@ -0,0 +1,209 @@
+
+    /**
+     * Stores the id of splits that has been assigned. The split assigner may be rebuilt when a
+     * DynamicFilteringEvent is received. After that, the splits that are already assigned can be
+     * assigned for the second time. We have to retain the state and filter out the splits that has
+     * been assigned with this set.
+     */
+    private final Set<String> assignedSplits;
+
+    private transient Set<String> allEnumeratingSplits;
+
+    private transient FileSplitAssigner splitAssigner;
+
+    // ------------------------------------------------------------------------
+
+    public DynamicFileSplitEnumerator(
+            SplitEnumeratorContext<SplitT> context,
+            DynamicFileEnumerator.Provider fileEnumeratorFactory,
+            FileSplitAssigner.Provider splitAssignerFactory) {
+        this.context = checkNotNull(context);
+        this.splitAssignerFactory = checkNotNull(splitAssignerFactory);
+        this.fileEnumeratorFactory = checkNotNull(fileEnumeratorFactory);
+        this.assignedSplits = new HashSet<>();
+    }
+
+    @Override
+    public void start() {
+        // no resources to start
+    }
+
+    @Override
+    public void close() throws IOException {
+        // no resources to close
+    }
+
+    @Override
+    public void addReader(int subtaskId) {
+        // this source is purely lazy-pull-based, nothing to do upon registration
+    }
+
+    @Override
+    public void handleSplitRequest(int subtask, @Nullable String hostname) {
+        if (!context.registeredReaders().containsKey(subtask)) {
+            // reader failed between sending the request and now. skip this request.
+            return;
+        }
+
+        if (splitAssigner == null) {
+            // No DynamicFilteringData is received before the first split request,
+            // create a split assigner that handles all splits
+            createSplitAssigner(null);
+        }
+
+        if (LOG.isDebugEnabled()) {
+            final String hostInfo =
+                    hostname == null ? "(no host locality info)" : "(on host '" + hostname + "')";
+            LOG.debug("Subtask {} {} is requesting a file source split", subtask, hostInfo);
+        }
+
+        final Optional<FileSourceSplit> nextSplit = getNextUnassignedSplit(hostname);
+        if (nextSplit.isPresent()) {
+            final FileSourceSplit split = nextSplit.get();
+            context.assignSplit((SplitT) split, subtask);
+            assignedSplits.add(split.splitId());
+            LOG.debug("Assigned split to subtask {} : {}", subtask, split);
+        } else {
+            context.signalNoMoreSplits(subtask);
+            LOG.info("No more splits available for subtask {}", subtask);
+        }
+    }
+
+    private Optional<FileSourceSplit> getNextUnassignedSplit(String hostname) {
+        Optional<FileSourceSplit> nextSplit = splitAssigner.getNext(hostname);
+        while (nextSplit.isPresent()) {
```

Comment:
I'm a bit tend to we not keep a set of assignedSplits in this class, and remove the assigned ones on-the-fly. 

Currently `splitAssigner` has kept the information of assigned splits, which indicating some kind of information duplication, also the iteration here might cause a bit more overhead. 

Instead, we might
1. Not keep the assignedSplits set.
2. When creating a new splitAssigner, we could retrieve the unassigned sets from the existing splitAssigner. Then we could remove the splits not in the unassigned sets.

Reply:
I'm afraid we still need the assigned splits here. Suppose that the enumerator keeps split 1,2, and filtered out split 3, and 1 has been assigned. Then it receives an event without any filtering. It's now enumerating 1,2,3 again, but it isn't aware whether 1 and 3 are assigned, since the assigner only tells that 2 is not assigned yet.
The case should be handled by the DynamicFilteringDataCollectorCoordinator now by not sending duplicate events. But in case any further changes on that coordinator, and giving the fact that the duplication costs little memory, maybe it's ok to keep the set.